src/Conversion/ONNXToKrnl/ONNXToKrnlCommon.cpp

```diff
@@ -369,3 +369,170 @@ Value getDimOrConstant(ConversionPatternRewriter &rewriter, Location loc,
   return dimVal;
 }
+
+/// A helper fucntion to check whether a value is produced by a dense
+/// ONNXConstantOp.
+bool isDenseONNXConstant(Value result) {
+  Operation *op = result.getDefiningOp();
+
+  ONNXConstantOp constOp = llvm::dyn_cast_or_null<ONNXConstantOp>(op);
+  // Not a constant.
+  if (!constOp)
+    return false;
+
+  // If the dense attribute is null, there must be buffer_id
+  // attribute.
+  if (!(op->getAttrOfType<::mlir::Attribute>("value")))
+    return false;
+  // The other attributes must be null.
+  if (op->getAttrOfType<::mlir::Attribute>("sparse_value"))
+    return false;
+  if (op->getAttrOfType<::mlir::Attribute>("value_float"))
+    return false;
+  if (op->getAttrOfType<::mlir::Attribute>("value_floats"))
+    return false;
+  if (op->getAttrOfType<::mlir::Attribute>("value_int"))
+    return false;
+  if (op->getAttrOfType<::mlir::Attribute>("value_ints"))
+    return false;
+  if (op->getAttrOfType<::mlir::Attribute>("value_string"))
+    return false;
+  if (op->getAttrOfType<::mlir::Attribute>("value_strings"))
+    return false;
+
+  return true;
+}
+
+bool isKrnlGlobalConstant(Value result) {
+  Operation *op = result.getDefiningOp();
+
+  KrnlGlobalOp constOp = llvm::dyn_cast_or_null<KrnlGlobalOp>(op);
+  // Not a constant.
+  if (!constOp)
+    return false;
+
+  if (!(op->getAttrOfType<::mlir::Attribute>("value")))
+    return false;
+
+  return true;
+}
+
+Value emitSqueeze(ConversionPatternRewriter &rewriter, Location loc,
+    Type resultType, Value input, int64_t axis) {
+  if (isKrnlGlobalConstant(input) || isDenseONNXConstant(input)) {
+    char *inputBuffer = createArrayFromDenseElementsAttr(
+        input.getDefiningOp()
+            ->getAttrOfType<::mlir::Attribute>("value")
+            .dyn_cast_or_null<mlir::DenseElementsAttr>());
+
+    Value constVal = createDenseONNXConstantOp(
+        rewriter, loc, resultType.cast<ShapedType>(), inputBuffer)
+                         .getResult();
+    free(inputBuffer);
+    return constVal;
+  } else {
+    return rewriter
+        .create<ONNXSqueezeOp>(
+            loc, resultType, input, rewriter.getI64ArrayAttr(axis))
+        .getResult();
+  }
+}
+
+Value emitUnsqueeze(ConversionPatternRewriter &rewriter, Location loc,
+    Type resultType, Value input, int64_t axis) {
+  if (isKrnlGlobalConstant(input) || isDenseONNXConstant(input)) {
+    char *inputBuffer = createArrayFromDenseElementsAttr(
+        input.getDefiningOp()
+            ->getAttrOfType<::mlir::Attribute>("value")
+            .dyn_cast_or_null<mlir::DenseElementsAttr>());
+
+    Value constVal = createDenseONNXConstantOp(
+        rewriter, loc, resultType.cast<ShapedType>(), inputBuffer)
+                         .getResult();
+    free(inputBuffer);
+    return constVal;
+  } else {
+    return rewriter
+        .create<ONNXUnsqueezeOp>(
+            loc, resultType, input, rewriter.getI64ArrayAttr(axis))
+        .getResult();
+  }
+}
+
+// Only support evenly splitting.
```

Comment:
Could you just copy over the header file comment here too.

maybe the function name could be "splitConstantOrEmitSplitONNXOp"

Same comments for other operations below...

Reply:
Done.